tasmota/tasmota_xnrg_energy/xnrg_10_sdm630.ino

```diff
@@ -95,117 +92,72 @@ void SDM630Every250ms(void)
+
+          //0x000C
+          Energy->active_power[0] = convBufToFloat(&buffer[27]);      //  +   -   +   W    Phase 1 power
+          Energy->active_power[1] = convBufToFloat(&buffer[31]);      //  +   -   -   W    Phase 2 power
+          Energy->active_power[2] = convBufToFloat(&buffer[35]);      //  +   -   -   W    Phase 3 power
+
+          //0x0012
+          Energy->apparent_power[0] = convBufToFloat(&buffer[39]);    //  +   -   +   VA   Phase 1 volt amps
+          Energy->apparent_power[1] = convBufToFloat(&buffer[43]);    //  +   -   -   VA   Phase 2 volt amps
+          Energy->apparent_power[2] = convBufToFloat(&buffer[47]);    //  +   -   -   VA   Phase 3 volt amps
+
+          //0x0018
+          Energy->reactive_power[0] = convBufToFloat(&buffer[51]);    //  +   -   +   VAr  Phase 1 volt amps reactive
+          Energy->reactive_power[1] = convBufToFloat(&buffer[55]);    //  +   -   -   VAr  Phase 2 volt amps reactive
+          Energy->reactive_power[2] = convBufToFloat(&buffer[59]);    //  +   -   -   VAr  Phase 3 volt amps reactive
+
+          //0x001E
+          Energy->power_factor[0] = convBufToFloat(&buffer[63]);      //  +   -   +        Phase 1 power factor
+          Energy->power_factor[1] = convBufToFloat(&buffer[67]);      //  +   -   -        Phase 2 power factor
+          Energy->power_factor[2] = convBufToFloat(&buffer[71]);      //  +   -   -        Phase 3 power factor
+          break;
+
+        case 1: // start address 0x0046
+          Energy->frequency[0] = convBufToFloat(&buffer[3]);          //  +   +   +   Hz   Frequency of supply voltages
+          break;
+
+        case 2: // start address 0x0156
+          //   0x0156   //  +   +   +   kWh  Total active energy
+          //   0x0158   //  +   +   +   kvarh Total reactive energy
+
+#ifdef SDM630_IMPORT
+          //0x015A
+          Energy->import_active[0] = convBufToFloat(&buffer[11]);     //  +   +   +   kWh  Phase 1 import active energy
+          Energy->import_active[1] = convBufToFloat(&buffer[15]);     //  +   +   +   kWh  Phase 2 import active energy
+          Energy->import_active[2] = convBufToFloat(&buffer[19]);     //  +   +   +   kWh  Phase 3 import active energy
+#endif
+          //0x0160
+          Energy->export_active[0] = convBufToFloat(&buffer[23]);     //  +   +   +   kWh  Phase 1 export active energy
+          Energy->export_active[1] = convBufToFloat(&buffer[27]);     //  +   +   +   kWh  Phase 2 export active energy
+          Energy->export_active[2] = convBufToFloat(&buffer[31]);     //  +   +   +   kWh  Phase 3 export active energy
 
-        case 22:
-//          Energy->import_active[0] = value;
           EnergyUpdateTotal();
           break;
       }
 
       Sdm630.read_state++;
-      if (sizeof(sdm630_start_addresses)/2 == Sdm630.read_state) {
+      if ( (sizeof(sdm630ReqConf)/sizeof(sdm630ReqConf[0])) == Sdm630.read_state) {
```

Comment:
Just wondering why we do not habe a macro `NELEM` for this. Saw this on many POSIX systems.

Reply:
Well I have a macro called ``nitems`` which does just that ;-)